### PR TITLE
Enables ARPACK, restructures some code.

### PR DESCRIPTION
**Pull request by cpviolator**
This branch enables the ARPACK linear solver BLAS library, and includes a function that will calculate the eigenpairs of the full Dirac operator, then project them out. The remainder is calculated stochastically with an MG inverter.

### CODE REVIEW

**Comment by cpviolator** [comment]
@ckallidonis @khadjiyiannakou Can you guys review and approve this pull? Once merged, further additions can be added to the master branch.

File: README
Comment (ckallidonis) [inline README:77]: As it is stated here it might imply that Kyriakos and I have developed the whole QUDA library. I would suggest that we are a bit more specific here. We could clearly state that Kyriakos and I developed the QKXTM code parts, e.g.:

The QKXTM code pieces were developed by,

Kyriakos Hadjiyiannakou
Christos Kallidonis

File: lib/qudaQKXTM_Kepler_utils.cpp
Comment (ckallidonis) [inline lib/qudaQKXTM_Kepler_utils.cpp:1]: I find it a very good idea to split the qudaQKXTM_Kepler.cpp into separate files. In this sense, I have the following suggestions:

- Create a new file lib/qudaQKXTM_Loop_Kepler.cpp. Move functions:
    - oneEndTrick_w_One_Der from qudaQKXTM_Kepler_utils.cpp:line-212 into qudaQKXTM_Loop_Kepler.cpp
    - writeLoops_ASCII from from qudaQKXTM_Kepler_utils.cpp:line-867 into qudaQKXTM_Loop_Kepler.cpp
    - getLoopWriteBuf from qudaQKXTM_Kepler_utils.cpp:line-941 into qudaQKXTM_Loop_Kepler.cpp
    - writeLoops_HDF5 from qudaQKXTM_Kepler_utils.cpp:line-967 into qudaQKXTM_Loop_Kepler.cpp
    - Loop_w_One_Der_FullOp_Exact from qudaQKXTM_Deflation_Kepler.cpp:line-1916 into qudaQKXTM_Loop_Kepler.cpp

With this, we also gather the loop-related functions into their own separate file, and qudaQKXTM_Kepler_utils.cpp is left with just the real "utility functions".

File: qkxtm/QKXTM_util.cpp
Comment (ckallidonis) [inline qkxtm/QKXTM_util.cpp:1619]: I would suggest to change the default smethod to "smethod = 0", as it's the one we have been using so far.

**Review by ckallidonis — COMMENTED** [review]
As a general comment, we could also move the commented functions at the end of their corresponding files for better convenience when navigating through the files.

**Comment by cpviolator** [comment]
Great suggestions! Having a separate loop file makes a lot of sense. I'll implement these soon and @ you guys when they are ready for review.

**Comment by cpviolator** [comment]
Aslo, it might be a good idea to include all the old functions (deflateAndInvert, etc) not as commented out code, but under a preprocessor flag such as DEPRECATED or QKXTM_DEBUG. That way, one can choose to compile the code via CMake, and the context highlighting for the code will still show up.

**Comment by ckallidonis** [comment]
I am not quite sure if this is going to be smooth, as many of the deprecated functions are really out-of-date, and if we uncomment them, we might need to update them to at least be able to compile.

**Comment by cpviolator** [comment]
OK, good call. These functions have clearly been well tested in the past.

**Review by khadjiyiannakou — APPROVED** [review]

**Comment by cpviolator** [comment]
I've pushed some changed to mainline QUDA that will help us separate out this code further. https://github.com/lattice/quda/pull/565/files
Mainline QUDA now has two extra auxiliary preconditioner pointers:
```c++
/** Preconditioner instance, e.g., multigrid */
void *preconditioner;
void *preconditionerUP;
void *preconditionerDN;
```
and `double delta_*MG;` variables
```c++
/**< Experimental MG additions */
double delta_muMG;
double delta_kappaMG;    
double delta_massMG;
```
with appropriate changes made to the `mutigrid_solver` function and `multigrid_invert_test.cpp` file.

I'm going to work out how to use a precompiled version of mainline QUDA to compile QKXTM. This involves using the `.o` files directly, rather than recompiling all kernels from scratch. This has the desirable benefit that we do not have to host mainline QUDA code, and upgrading QKXTM will be reduced to `CMakeLists.txt` files only, greatly minimising maintenance overhead.

**Pull request merged**
